src/lib/core/fiber.c

```diff
@@ -903,7 +910,7 @@ fiber_recycle(struct fiber *fiber)
 	if (!has_custom_stack) {
 		rlist_move_entry(&cord()->dead, fiber, link);
 	} else {
-		fiber_destroy(cord(), fiber);
+		cord_add_garbage(cord(), fiber);
```

Comment:
Instead of introducing the garbage fiber in addition to dead fibers, can't we simply delete the stack for such a fiber, add it to the list of dead fibers like a typical fiber, and allocate a stack of the proper size when the fiber gets recycled?

Reply:
We can't delete the stack right away. See the comment in `cord.garbage` member:
```C
	/**
	 * Latest dead fiber which couldn't be reused and waits for its
	 * deletion. A fiber can't be reused if it is somehow non-standard. For
	 * instance, has a special stack.
	 * A fiber can't be deleted if it is the current fiber - can't delete
	 * own stack safely. Then it schedules own deletion for later. The
	 * approach is very similar to pthread stacks deletion - pthread can't
	 * delete own stack, so they are saved and deleted later by a newer
	 * pthread or by some other dying pthread. Same here with fibers.
	 */
	struct fiber *garbage;
```
In particular: *A fiber can't be deleted if it is the current fiber - can't delete own stack safely*. You won't even be able to return from `fiber_recycle` if you free the stack you are standing on.

As for why not to add the fiber itself to the list of recycled fibers - but how is it better than putting it back into the mempool? You will need to handle fibers both having and not having a stack in the fiber cache. Now the cache contains only normal fibers having usual-sized stack.

Reply:
> As for why not to add the fiber itself to the list of recycled fibers - but how is it better than putting it back into the mempool? You will need to handle fibers both having and not having a stack in the fiber cache. Now the cache contains only normal fibers having usual-sized stack.

Agree, that's probably a bad idea, because this could slow down creation of normal fibers.

Reply:
What about adding a dead fiber with a custom stack to the `cord->dead` list and deleting it in `fiber_new_ex()` instead of reusing it? Pros: no need for the new `cord->garbage` entity.

Another suggestion: can we delete the `cord->garbage` fiber in just one place, in `fiber_yield()` after `coro_swtich()`?

Reply:
We certainly can do both. However:

1 - Adding to `cord->dead` list would mean the list stores both fibers for deletion and for reuse. I considered it at first but didn't want to complicate handling of `dead` list. Pros is indeed we won't need `garbage` member, but cons - `dead` list will contain both garbage and cached fibers. If you want, I can do that. Let me know.

2 - I considered `fiber_yield()` as well. But 2 problems: firstly, you might never return to this fiber again after `fiber_yield()`. Then it would leak. Secondly, you would return again onto the same stack which you want to delete. Hence deletion would be unsafe again, like before `fiber_yield()`.

You probably meant that we delete the previous fiber in the next fiber's context to which we switched. Then it is possible. Although that would add +1 `if` to a hot path - on each yield. To check if `cord->garbage` is not `NULL` or to check if the first fiber in `cord->dead` is garbage. That might affect perf, I am afraid. But I didn't check of course. Anyway, now the `if` is added only to fiber creation and deletion, which should be much rarer than yields.

Reply:
1. Mixing reusable and useless fibers on the same list is probably a bad idea, you're right.
2. Adding extra code to the hot path is bad.

Let's leave everything as is.